### PR TITLE
Remove hard dependency on pandas in get_dummies

### PR DESCRIPTION
**Pull request by galipremsagar**
- [ ] Tests added / passed
- [x] Passes `black dask` / `flake8 dask`

This fixes: #4885

### CODE REVIEW

**Comment by galipremsagar** [comment]
rerun tests

File: dask/dataframe/reshape.py [hunk 1]
```diff
@@ -106,16 +107,17 @@ def get_dummies(
     """
     if PANDAS_VERSION >= "0.23.0":
         # dtype added to pandas
-        kwargs = {"dtype": dtype}
+        if len(kwargs) == 0:
+            kwargs = {"dtype": dtype}
```
Comment (mrocklin) [inline dask/dataframe/reshape.py]: Why this change?

File: dask/dataframe/reshape.py [hunk 2]
```diff
@@ -161,19 +163,33 @@ def get_dummies(
...
+            data._meta,
+            prefix=prefix,
+            prefix_sep=prefix_sep,
+            dummy_na=dummy_na,
+            columns=columns,
+            sparse=sparse,
+            drop_first=drop_first,
+            **kwargs
+        )
+        method_cache = M
```
Comment (mrocklin) [inline dask/dataframe/reshape.py]: A few questions/comments here:

1.  It seems odd to special case Pandas here.  Ideally we would find a single solution that would work for any pandas-like library
2.  I don't understand the use of `M` here.  The `M` object creates methods, so for example this would be equivalend to `data._meta.get_dummies`.  My understanding is that this doesn't exist for Pandas.

@TomAugspurger , do you have any suggestions for how to get a function or a module of an object to go from a Pandas dataframe to a Pandas function (not method).

File: dask/dataframe/reshape.py [hunk 3]
```diff
@@ -161,19 +163,33 @@ def get_dummies(
...
+            data._meta,
+            prefix=prefix,
+            prefix_sep=prefix_sep,
+            dummy_na=dummy_na,
+            columns=columns,
+            sparse=sparse,
+            drop_first=drop_first,
+            **kwargs
+        )
+        method_cache = pd
```
Comment (mrocklin) [inline dask/dataframe/reshape.py]: I don't understand the use of the term "method cache" for this.

Comment (galipremsagar) [reply dask/dataframe/reshape.py]: This should go away if I have a better way to call get_dummies rather than from Method cache. But my understanding was that M was a Method cache for package functions. But I'm wrong and it doesn't seem to be the case.

File: dask/dataframe/reshape.py (hunk 2 above)
Comment (galipremsagar) [reply dask/dataframe/reshape.py]: True, because of point 2 I had to have a branched condition for pandas separately. But I would like to know if we can do it in a more common code like you asked @TomAugspurger. I'm aswell looking to see how I can find package level function as both pandas & cudf now have `get_dummies` as a package level functions.

File: dask/dataframe/reshape.py (hunk 1 above)
Comment (galipremsagar) [reply dask/dataframe/reshape.py]: Because of the reason I had to introduce `kwargs` in `get_dummies` call. I had to do this as currently cudf's get_dummies behavior is slightly different from that of panda's. For the `meta` to be populated for dask to work we would require `cats` attribute to be passed for `cudf.get_dummies` to give the expected columns on an empty dataframe. So instead of creating inconsistency by adding `cats` to dask get_dummies call, moved the kwargs being created below to the function signature itself and having an equivalent check which was being done previously.

File: dask/dataframe/reshape.py (hunk 2 above)
Comment (TomAugspurger) [reply dask/dataframe/reshape.py]: > Pandas dataframe to a Pandas function (not method).
 
Sorry, I don't quite follow. Is it fair to say that in e.g. `dask.dataframe.get_dummies`, we'd like to get at `DataFrame.get_dummies`? Does `getattr(data, <name>)` work?

Comment (mrocklin) [reply dask/dataframe/reshape.py]: Unfortuantely not in this case

```python
In [1]: import pandas as pd

In [2]: df = pd.DataFrame()

In [3]: df.get_dummies
AttributeError: 'DataFrame' object has no attribute 'get_dummies'

In [4]: pd.get_dummies
Out[4]: <function pandas.core.reshape.reshape.get_dummies(data, prefix=None, prefix_sep='_', dummy_na=False, columns=None, sparse=False, drop_first=False, dtype=None)>
```

Comment (galipremsagar) [reply dask/dataframe/reshape.py]: @TomAugspurger, like Matthew mentioned I'm trying to access package level function since `get_dummies` is not a DataFrame function in pandas. 
```python
>>> import cudf
>>> cudf.get_dummies
<function get_dummies at 0x7f0b55f63ae8>

```

Comment (TomAugspurger) [reply dask/dataframe/reshape.py]: Ah understood.

Is this too hacky?

```
In [34]: obj = pd.DataFrame()

In [35]: package = obj.__class__.__module__.split(".")[0]

In [36]: sys.modules[package].get_dummies
Out[36]: <function pandas.core.reshape.reshape.get_dummies(data, prefix=None, prefix_sep='_', dummy_na=False, columns=None, sparse=False, drop_first=False, dtype=None)>
```

Comment (mrocklin) [reply dask/dataframe/reshape.py]: It's certainly hacky :)  But also pragmatic :/

I think I'll leave this question to you if you don't mind :)

Comment (TomAugspurger) [reply dask/dataframe/reshape.py]: I can live with it. I don't think it's likely to break (famous last words).

Comment (galipremsagar) [reply dask/dataframe/reshape.py]: Made changes accordingly.

File: dask/dataframe/reshape.py (hunk 1 above)
Comment (mrocklin) [reply dask/dataframe/reshape.py]: I see, so you're saying that there is an extra keyword argument that cudf wants, so you want to include `**kwargs` in the dask.dataframe.get_dummies signature?  This makes sense.  

Given that though, I think that you can probably replace these four lines as follows?:

```diff
-if len(kwargs) == 0:
-    kwargs = {'dtype': dtype}
-else:
-    kwargs['dtype'] = dtype
+kwargs['dtype'] = dtype

Comment (galipremsagar) [reply dask/dataframe/reshape.py]: Yes, since `kwargs` would be an empty dict it makes sense to remove the length check. Removed it. 👍

**Comment by mrocklin** [comment]
This looks good to me.  Thanks for the changes and the effort here @galipremsagar .  Merging.
